### PR TITLE
 implementation of Scaffold and FedProx for TensorFlow

### PR DESCRIPTION
**Pull request by falibabaei**
###  implementation of Scaffold and FedProx for TensorFlow

Here's the implementation of Scaffold and FedProx for TensorFlow. We tested our code with TensorFlow versions 2.10 up to 2.17. Two scripts,  `fedprox_loss.py ` and  `scaffold.py `,   have been added to  `nvflare/app_opt/tf ` to calculate the FedProx and SCAFFOLD term updates, respectively. 

If the data distribution is highly uneven, issues may arise with the SCAFFOLD implementation in TensorFlow. These problems typically become apparent only after many training rounds and epochs. At some point, convergence issues may occur, and model weights may explode, leading to NaN values in weights and losses. This problem was observed specifically with a distribution where α=0.1. To address this issue, a clipping norm was applied.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [x] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

**Comment by falibabaei** [comment]
@chesterxgchen Sorry but it was related to the step of updating the documentation in the [contributing guidelines](https://github.com/NVIDIA/NVFlare/blob/main/CONTRIBUTING.md). We have updated it.

File: docs/apidocs/modules.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+nvflare
+=======
+
+.. toctree::
+   :maxdepth: 4
+
+   nvflare
```
Comment (YuanTingHsieh) [inline docs/apidocs/modules.rst:7]: do we need this? @nvkevlu @SYangster

File: examples/getting_started/pt/src/cifar10_fl.py [hunk 2]
```diff
@@ -122,6 +122,7 @@ def evaluate(input_weights):
 
         # (6) evaluate on received model for model selection
         accuracy = evaluate(input_model.params)
+        summary_writer.add_scalar(tag="global_model_accuracy", scalar=accuracy, global_step=input_model.current_round)
```
Comment (YuanTingHsieh) [inline examples/getting_started/pt/src/cifar10_fl.py:125]: if you think this is necessary, could you move this change to another PR?
So this PR only relates to TF changes, thanks

File: examples/getting_started/tf/src/cifar10_tf_fl.py
Comment (YuanTingHsieh) [inline examples/getting_started/tf/src/cifar10_tf_fl.py:1]: this file is used in other notebooks, please add it back.

File: integration/nemo/examples/peft/nemo_nvflare/megatron_gpt_peft_tuning.py [hunk 3]
```diff
@@ -13,15 +13,13 @@
 # limitations under the License.
 
 import torch.multiprocessing as mp
-from omegaconf.omegaconf import OmegaConf
```
Comment (YuanTingHsieh) [inline integration/nemo/examples/peft/nemo_nvflare/megatron_gpt_peft_tuning.py:16]: please revert changes to this file

File: examples/getting_started/tf/tf_fedavg_script_executor_cifar10.py
Comment (YuanTingHsieh) [inline examples/getting_started/tf/tf_fedavg_script_executor_cifar10.py:1]: @holgerroth please check if we still need this file, thanks

**Review by YuanTingHsieh — COMMENTED** [review]
@falibabaei thanks for contributing back!

Great examples with code, instructions, and results for the TF users.
Please ensure that changes are confined to the files specified in this PR. Do not modify files that are outside the scope of this PR.

@holgerroth please help take a look as well, thanks!

**Comment by LeoDuda** [comment]
@YuanTingHsieh,

our changes just involve `nvflare/app_opt/tf/scaffold.py` and some of the files within `examples/getting_started/tf`.
Besides that, no changes where made by us. We forked https://github.com/holgerroth/NVFlare/tree/tf_algos and other changes stated above should be related to this work.
We used this branch because of the the newly introduced workflow and a working implementation of FedOpt.

File: examples/getting_started/pt/src/cifar10_fl.py (hunk 2 above)
Comment (LeoDuda) [reply examples/getting_started/pt/src/cifar10_fl.py:125]: @YuanTingHsieh,

our changes just involve nvflare/app_opt/tf/scaffold.py and some of the files within examples/getting_started/tf.
Besides that, no changes where made by us. We forked https://github.com/holgerroth/NVFlare/tree/tf_algos and other changes stated above should be related to this work.
We used this branch because of the the newly introduced workflow and a working implementation of FedOpt.

**Comment by YuanTingHsieh** [comment]
@falibabaei @LeoDuda since #2704 is merged, please rebase this PR on top of main and resolving any conflicts.

Thank you!

File: integration/xgboost/encryption_plugins/tests/test_tenseal.py [hunk 4]
```diff
@@ -95,12 +96,12 @@ def test_hori() -> None:
         out_len1 = ctypes.c_size_t()
 
         nvflare.FederatedPluginSyncEnrcyptedHistHori(
-                handle,
-                out,
-                out_len,
-                ctypes.byref(out1),
-                ctypes.byref(out_len1),
-            )
+            handle,
```
Comment (YuanTingHsieh) [inline integration/xgboost/encryption_plugins/tests/test_tenseal.py:99]: Style fixes in other places are great, but please move these changes into another PR.

File: nvflight/build_wheel.py [hunk 5]
```diff
@@ -34,13 +34,14 @@
 
 def patch(setup_dir, patch_file):
     file_dir_path = os.path.abspath(os.path.dirname(__file__))
-    cmd = ['git', 'apply', os.path.join(file_dir_path, patch_file)]
+    cmd = ["git", "apply", os.path.join(file_dir_path, patch_file)]
```
Comment (YuanTingHsieh) [inline nvflight/build_wheel.py:37]: Style fix of unrelated files should be in another PR.

File: nvflight/prepare_setup.py [hunk 6]
```diff
@@ -19,88 +19,41 @@
 exclude_extensions = [".md", ".rst", ".pyc", "__pycache__"]
 
 nvflight_packages = {
```
Comment (YuanTingHsieh) [inline nvflight/prepare_setup.py:21]: Style fix of unrelated files should be in another PR.

File: nvflight/setup.py [hunk 7]
```diff
@@ -32,7 +32,9 @@
     exclude=["tests", "tests.*"],
 )
 
-package_data = {"": ["*.yml", "*.config"], }
```
Comment (YuanTingHsieh) [inline nvflight/setup.py:35]: Style fix of unrelated files should be in another PR.

File: setup.py [hunk 8]
```diff
@@ -100,4 +100,3 @@ def remove_dir(target_path):
 )
```
Comment (YuanTingHsieh) [inline setup.py:101]: Style fix of unrelated files should be in another PR.

File: versioneer.py
Comment (YuanTingHsieh) [inline versioneer.py:1]: Style fix of unrelated files should be in another PR.

Comment (falibabaei) [reply versioneer.py:1]: Done

File: versioneer.py [hunk 9]
```diff
@@ -2106,4 +2106,4 @@ def scan_setup_py():
         errors = do_setup()
         errors += scan_setup_py()
         if errors:
-            sys.exit(1)
+            sys.exit(1)
```
Comment (YuanTingHsieh) [inline versioneer.py:2109]: Style fix of unrelated files should be in another PR.

File: nvflight/setup.py [hunk 10]
```diff
@@ -51,4 +51,4 @@
     packages=release_package,
     package_data=package_data,
     include_package_data=True,
-)
+)
```
Comment (YuanTingHsieh) [inline nvflight/setup.py:54]: please just undo this change, thanks

File: nvflight/prepare_setup.py [hunk 11]
```diff
@@ -183,4 +183,4 @@ def prepare_setup(setup_dir: str):
 
     for src in src_files:
         shutil.copy(src, os.path.join(setup_dir, os.path.basename(src)))
-    
+
```
Comment (YuanTingHsieh) [inline nvflight/prepare_setup.py:186]: please just undo this change, thanks

File: nvflight/build_wheel.py [hunk 12]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (YuanTingHsieh) [inline nvflight/build_wheel.py:1]: please just undo this change, thanks

File: nvflare/app_opt/tf/scaffold.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (YuanTingHsieh) [inline nvflare/app_opt/tf/scaffold.py:1]: change to 2024

File: examples/getting_started/tf/src/cifar10_tf_fl_alpha_split_scaffold.py [hunk 14]
```diff
@@ -0,0 +1,294 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (YuanTingHsieh) [inline examples/getting_started/tf/src/cifar10_tf_fl_alpha_split_scaffold.py:1]: should be 2024

**Review by YuanTingHsieh — COMMENTED** [review]
mostly good, suggest undo non-required changes

File: nvflight/build_wheel.py (hunk 12 above)
Comment (falibabaei) [reply nvflight/build_wheel.py]: Done. Thanks

File: docs/apidocs/modules.rst (hunk 1 above)
Comment (SYangster) [reply docs/apidocs/modules.rst:7]: This change isn't needed we should remove this. The apidocs are automatically generated when building the documentation.

Comment (YuanTingHsieh) [reply docs/apidocs/modules.rst:7]: @falibabaei can you undo this one as well? Thanks

Comment (falibabaei) [reply docs/apidocs/modules.rst]: Done

**Review by YuanTingHsieh — APPROVED** [review]

**Comment by YuanTingHsieh** [comment]
@falibabaei @LeoDuda next time please try to rebase (squash commits) and sign all the commits thanks!
